url47e6aca666_pocorall_automaton_tgz-pJ8-net_pocorall_automaton_DatatypesJ8/src/net/pocorall/automaton/BasicOperations.java

```diff
@@ -0,0 +1,622 @@
+	 * <p/>
+	 * Complexity: linear in number of states.
+	 */
+	public static DefaultAutomaton union(Collection<DefaultAutomaton> l) {
+		Set<Integer> ids = new HashSet<Integer>();
+		for (DefaultAutomaton a : l)
+			ids.add(System.identityHashCode(a));
+		boolean has_aliases = ids.size() != l.size();
+		State s = new State();
+		for (DefaultAutomaton b : l) {
+			if (b.isEmpty())
+				continue;
+			DefaultAutomaton bb = b;
+			if (has_aliases)
+				bb = bb.cloneExpanded();
+			else
+				bb = bb.cloneExpandedIfRequired();
+			s.addEpsilon(bb.initial);
+		}
+		DefaultAutomaton a = new DefaultAutomaton();
+		a.initial = s;
+		a.deterministic = false;
+		a.clearHashCode();
+		a.checkMinimizeAlways();
+		return a;
+	}
+
+	/**
+	 * Determinizes the given automaton.
+	 * <p/>
+	 * Complexity: exponential in number of states.
+	 */
+	public static void determinize(LinkedAutomaton a) {
+		if (a.deterministic)
+			return;
+		Set<State> initialset = new HashSet<State>();
+		initialset.add(a.initial);
+		determinize(a, initialset);
+	}
+
+	/**
+	 * Determinizes the given automaton using the given set of initial states.
+	 */
+	static void determinize(LinkedAutomaton a, Set<State> initialset) {
+		char[] points = a.getStartPoints();
+		// subset construction
+		Map<Set<State>, Set<State>> sets = new HashMap<Set<State>, Set<State>>();
+		LinkedList<Set<State>> worklist = new LinkedList<Set<State>>();
+		Map<Set<State>, State> newstate = new HashMap<Set<State>, State>();
+		sets.put(initialset, initialset);
+		worklist.add(initialset);
+		a.initial = new State();
+		newstate.put(initialset, a.initial);
+		while (worklist.size() > 0) {
+			Set<State> s = worklist.removeFirst();
+			State r = newstate.get(s);
+			for (State q : s)
+				if (q.accept != null) {
+					r.accept = true;
+					break;
+				}
+			for (int n = 0; n < points.length; n++) {
+				Set<State> p = new HashSet<State>();
+				for (State q : s)
+					for (Transition t : q.transitions)
+						if (t.min <= points[n] && points[n] <= t.max)
+							p.add(t.to);
+				if (!sets.containsKey(p)) {
+					sets.put(p, p);
+					worklist.add(p);
+					newstate.put(p, new State());
+				}
+				State q = newstate.get(p);
+				char min = points[n];
+				char max;
+				if (n + 1 < points.length)
+					max = (char) (points[n + 1] - 1);
+				else
+					max = Character.MAX_VALUE;
+				r.transitions.add(new Transition(min, max, q));
+			}
+		}
+		a.deterministic = true;
+		a.removeDeadTransitions();
+	}
+
+	/**
+	 * Adds epsilon transitions to the given automaton.
+	 * This method adds extra character interval transitions that are equivalent to the given
+	 * set of epsilon transitions.
+	 *
+	 * @param pairs collection of {@link StatePair} objects representing pairs of source/destination states
+	 *              where epsilon transitions should be added
+	 */
+	public static void addEpsilons(LinkedAutomaton a, Collection<StatePair> pairs) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 12. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code blocks on lines 500-513, lines 515-543 into separate methods.